include/nbl/builtin/hlsl/bxdf/fresnel.hlsl

```diff
@@ -172,58 +172,66 @@ struct ReflectRefract
     using vector_type = vector<T, 3>;
     using scalar_type = T;
 
-    static this_t create(bool r, NBL_CONST_REF_ARG(vector_type) I, NBL_CONST_REF_ARG(vector_type) N, scalar_type NdotI, scalar_type NdotTorR, scalar_type rcpOrientedEta)
+    static this_t create(NBL_CONST_REF_ARG(vector_type) I, NBL_CONST_REF_ARG(vector_type) N, scalar_type NdotI, scalar_type rcpOrientedEta)
     {
         this_t retval;
         retval.I = I;
         retval.N = N;
         retval.NdotI = NdotI;
-        retval.NdotTorR = NdotTorR;
         retval.rcpOrientedEta = rcpOrientedEta;
         return retval;
     }
 
-    static this_t create(bool r, NBL_CONST_REF_ARG(Refract<scalar_type>) refract, scalar_type rcpOrientedEta)
+    static this_t create(NBL_CONST_REF_ARG(Refract<scalar_type>) refract, scalar_type rcpOrientedEta)
     {
         this_t retval;
         retval.I = refract.I;
         retval.N = refract.N;
         retval.NdotI = refract.NdotI;
-        retval.NdotTorR = hlsl::mix(refract.NdotI, refract.NdotT, r);
         retval.rcpOrientedEta = rcpOrientedEta;
         return retval;
     }
 
     // when you know you'll reflect
     void recomputeNdotR()
     {
-        refract.recomputeNdotI();
+        _refract.recomputeNdotI();
+        NdotI = _refract.NdotI;
     }
 
     // when you know you'll refract
     void recomputeNdotT(bool backside, scalar_type _NdotI2, scalar_type rcpOrientedEta2)
     {
-        refract.recomputeNdotT(backside, _NdotI2, rcpOrientedEta2);
+        _refract.recomputeNdotT(backside, _NdotI2, rcpOrientedEta2);
+    }
+
+    vector_type reflect()
+    {
+        return N * NdotI * 2.0f - I;
+    }
+
+    vector_type refract()
+    {
+        return N * (NdotI * rcpOrientedEta + _refract.NdotT) - I * rcpOrientedEta;
     }
 
     vector_type operator()(const bool doRefract)
     {
-        return N * (NdotI * (hlsl::mix<scalar_type>(1.0f, rcpOrientedEta, doRefract)) + NdotTorR) - I * (hlsl::mix<scalar_type>(1.0f, rcpOrientedEta, doRefract));
+        return N * (NdotI * (hlsl::mix<scalar_type>(1.0f, rcpOrientedEta, doRefract)) + hlsl::mix(NdotI, _refract.NdotT, doRefract)) - I * (hlsl::mix<scalar_type>(1.0f, rcpOrientedEta, doRefract));
```

Comment:
the `NdotTorR` should come from outside and not be computed as a member

having it as a member, getting recomputed, storing it gives you a really bad choice between:
1. no optimization (can't provide your own dot product value)
2. user-error prone (someone forgets to call `recompute` before `operator()`

https://github.com/Devsh-Graphics-Programming/Nabla/pull/899#discussion_r2197742248